File: app/src/main/java/com/battlelancer/seriesguide/ui/episodes/EpisodesAdapter.java

package com.battlelancer.seriesguide.ui.episodes;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.widget.TextViewCompat;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.databinding.ItemEpisodeBinding;
import com.battlelancer.seriesguide.provider.SgEpisode2Info;
import com.battlelancer.seriesguide.settings.DisplaySettings;
import com.battlelancer.seriesguide.util.TextTools;
import com.battlelancer.seriesguide.util.TimeTools;
import com.battlelancer.seriesguide.widgets.WatchedBox;
import com.uwetrottmann.androidutils.CheatSheet;
import java.text.NumberFormat;
import java.util.Date;
import java.util.List;

class EpisodesAdapter extends ArrayAdapter<SgEpisode2Info> {

    interface ClickListener {
        void onWatchedBoxClick(long episodeId, boolean isWatched);
        void onPopupMenuClick(@NonNull View v, long episodeId, int episodeNumber,
                long releaseTimeMs, int watchedFlag, boolean isCollected);
    }

    private final ClickListener clickListener;
    private final NumberFormat integerFormat;

    EpisodesAdapter(Context context, ClickListener listener) {
        super(context, 0);
        clickListener = listener;
        integerFormat = NumberFormat.getIntegerInstance();
    }

    void setData(List<SgEpisode2Info> data) {
        clear();
        if (data != null) {
            addAll(data);
        }
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    /**
     * Use episode ID instead of position to ensure Adapter.getItemIdAtPosition is stable.
     */
    @Override
    public long getItemId(int position) {
        // AbsListView tries to confirm checked positions regardless if count is 0, so check.
        if (position > getCount() - 1) {
            return -1;
        }
        SgEpisode2Info item = getItem(position);
        if (item != null) {
            return item.getId();
        } else {
            return -1;
        }
    }

    /**
     * Get the item position in the data set, or the position of the first item if it is not found.
     */
    int getItemPosition(long itemId) {
        for (int position = 0; position < getCount(); position++) {
            SgEpisode2Info item = getItem(position);
            if (item == null) {
                return 0;
            }
            if (item.getId() == itemId) {
                return position;
            }
        }
        return 0;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        ViewHolder viewHolder;

        if (convertView == null) {
            viewHolder = new ViewHolder(parent, clickListener, integerFormat);
        } else {
            viewHolder = (ViewHolder) convertView.getTag();
        }

        SgEpisode2Info item = getItem(position);
        if (item != null) {
            viewHolder.bindTo(item, getContext());
        }

        return viewHolder.binding.getRoot();
    }

    static class ViewHolder {
        private final ItemEpisodeBinding binding;
        private final ClickListener clickListener;
        private final NumberFormat integerFormat;

        public ViewHolder(ViewGroup parent,
                ClickListener clickListener,
                NumberFormat integerFormat) {
            this.clickListener = clickListener;
            this.integerFormat = integerFormat;
            this.binding = ItemEpisodeBinding
                    .inflate(LayoutInflater.from(parent.getContext()), parent, false);
            this.binding.getRoot().setTag(this);
        }

        public void bindTo(SgEpisode2Info episode, Context context) {
            // episode title
            final int watchedFlag = episode.getWatched();
            final int episodeNumber = episode.getEpisodenumber();
            boolean hideTitle = EpisodeTools.isUnwatched(watchedFlag)
                    && DisplaySettings.preventSpoilers(context);
            binding.textViewEpisodeTitle.setText(TextTools.getEpisodeTitle(context,
                    hideTitle ? null : episode.getTitle(), episodeNumber));

            // number
            binding.textViewEpisodeNumber.setText(integerFormat.format(episodeNumber));

            // watched box
            binding.watchedBoxEpisode.setEpisodeFlag(watchedFlag);
            binding.watchedBoxEpisode.setOnClickListener(v -> {
                WatchedBox box = (WatchedBox) v;
                // disable button, will be re-enabled on data reload once action completes
                box.setEnabled(false);
                clickListener.onWatchedBoxClick(episode.getId(),
                        !EpisodeTools.isWatched(box.getEpisodeFlag()));
            });
            binding.watchedBoxEpisode.setEnabled(true);
            boolean watched = EpisodeTools.isWatched(watchedFlag);
            binding.watchedBoxEpisode.setContentDescription(
                    context.getString(watched ? R.string.action_unwatched : R.string.action_watched));
            CheatSheet.setup(binding.watchedBoxEpisode,
                    watched ? R.string.action_unwatched : R.string.action_watched
            );

            // collected tag
            final boolean isCollected = episode.getCollected();
            binding.imageViewCollected.setVisibility(isCollected ? View.VISIBLE : View.INVISIBLE);

            // alternative numbers
            int absoluteNumber = episode.getAbsoluteNumber();
            String absoluteNumberText = null;
            if (absoluteNumber > 0) {
                absoluteNumberText = NumberFormat.getIntegerInstance().format(absoluteNumber);
            }
            double dvdNumber = episode.getDvdNumber();
            String dvdNumberText = null;
            if (dvdNumber > 0) {
                dvdNumberText = context.getString(R.string.episode_number_disk) + " " + dvdNumber;
            }
            binding.textViewEpisodeAlternativeNumbers.setText(
                    TextTools.dotSeparate(absoluteNumberText, dvdNumberText));

            // release time
            boolean isReleased;
            final long releaseTime = episode.getFirstReleasedMs();
            if (releaseTime != -1) {
                Date actualRelease = TimeTools.applyUserOffset(context, releaseTime);
                isReleased = TimeTools.isReleased(actualRelease);
                // "in 15 mins" or "Oct 31, 2010"
                boolean displayExactDate = DisplaySettings.isDisplayExactDate(context);
                binding.textViewEpisodeAirdate.setText(displayExactDate ?
                        TimeTools.formatToLocalDateShort(context, actualRelease)
                        : TimeTools.formatToLocalRelativeTime(context, actualRelease));
            } else {
                binding.textViewEpisodeAirdate.setText(context
                        .getString(R.string.episode_firstaired_unknown));
                isReleased = false;
            }

            // dim text color if not released
            TextViewCompat.setTextAppearance(binding.textViewEpisodeTitle, isReleased
                    ? R.style.TextAppearance_SeriesGuide_Subtitle1
                    : R.style.TextAppearance_SeriesGuide_Subtitle1_Dim);
            TextViewCompat.setTextAppearance(binding.textViewEpisodeAirdate, isReleased
                    ? R.style.TextAppearance_SeriesGuide_Body2_Secondary
                    : R.style.TextAppearance_SeriesGuide_Body2_Dim);

            // context menu
            this.binding.imageViewContextMenu.setOnClickListener(v -> clickListener
                    .onPopupMenuClick(v, episode.getId(), episodeNumber, releaseTime, watchedFlag,
                            isCollected));
        }
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/ui/episodes/EpisodesAdapter.kt
package com.battlelancer.seriesguide.ui.episodes

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.TooltipCompat
import androidx.core.widget.TextViewCompat
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.databinding.ItemEpisodeBinding
import com.battlelancer.seriesguide.provider.SgEpisode2Info
import com.battlelancer.seriesguide.settings.DisplaySettings.isDisplayExactDate
import com.battlelancer.seriesguide.settings.DisplaySettings.preventSpoilers
import com.battlelancer.seriesguide.util.TextTools
import com.battlelancer.seriesguide.util.TimeTools
import com.battlelancer.seriesguide.widgets.WatchedBox
import java.text.NumberFormat


class EpisodesAdapter(
    private val context: Context,
    private val clickListener: ClickListener
) : ListAdapter<SgEpisode2Info, EpisodeViewHolder>(SgEpisode2InfoDiffCallback) {

    var selectedItemId: Long = -1

    init {
        setHasStableIds(true)
    }

    // Is called (indirectly) by external code, so do size check.
    override fun getItemId(position: Int): Long =
        currentList.getOrNull(position)?.id ?: RecyclerView.NO_ID

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EpisodeViewHolder {
        return EpisodeViewHolder.create(parent, clickListener)
    }

    override fun onBindViewHolder(holder: EpisodeViewHolder, position: Int) {
        val item: SgEpisode2Info? = getItem(position)
        val isSelected = item?.id == selectedItemId
        holder.bind(item, isSelected, context)
    }

    /**
     * Returns -1 if the item ID was not found in the [getCurrentList].
     */
    fun getPositionForId(itemId: Long): Int {
        return currentList.indexOfFirst { it.id == itemId }
    }

    fun selectItem(position: Int): Long {
        val previousSelectedPosition = selectedItemId
            .let { if (it != -1L) getPositionForId(it) else -1 }
        val selectedItemId = getItemId(position)
            .also { selectedItemId = it }
        if (previousSelectedPosition != -1) notifyItemChanged(previousSelectedPosition)
        notifyItemChanged(position)
        return selectedItemId
    }

    interface ClickListener {
        fun onItemClick(position: Int)
        fun onWatchedBoxClick(episodeId: Long, isWatched: Boolean)
        fun onPopupMenuClick(
            v: View, episodeId: Long, episodeNumber: Int,
            releaseTimeMs: Long, watchedFlag: Int, isCollected: Boolean
        )
    }

}

object SgEpisode2InfoDiffCallback : DiffUtil.ItemCallback<SgEpisode2Info>() {
    override fun areItemsTheSame(oldItem: SgEpisode2Info, newItem: SgEpisode2Info): Boolean =
        oldItem.id == newItem.id

    override fun areContentsTheSame(oldItem: SgEpisode2Info, newItem: SgEpisode2Info): Boolean =
        oldItem == newItem
}

class EpisodeViewHolder(
    private val binding: ItemEpisodeBinding,
    clickListener: EpisodesAdapter.ClickListener
) : RecyclerView.ViewHolder(binding.root) {

    private val integerFormat = NumberFormat.getIntegerInstance()
    private var episode: SgEpisode2Info? = null

    init {
        binding.root.setOnClickListener {
            clickListener.onItemClick(absoluteAdapterPosition)
        }
        binding.watchedBoxEpisode.setOnClickListener { view ->
            episode?.let {
                val box = view as WatchedBox
                // disable button, will be re-enabled on data reload once action completes
                box.isEnabled = false
                clickListener.onWatchedBoxClick(it.id, !EpisodeTools.isWatched(box.episodeFlag))
            }
        }
        binding.imageViewContextMenu.setOnClickListener { v ->
            episode?.let {
                clickListener.onPopupMenuClick(
                    v,
                    it.id,
                    it.episodenumber,
                    it.firstReleasedMs,
                    it.watched,
                    it.collected
                )
            }
        }
    }

    fun bind(episode: SgEpisode2Info?, isActivated: Boolean, context: Context) {
        this.episode = episode

        // selection state
        binding.root.isActivated = isActivated

        if (episode == null) {
            binding.textViewEpisodeTitle.text = null
            binding.textViewEpisodeNumber.text = null
            binding.watchedBoxEpisode.episodeFlag = EpisodeFlags.UNWATCHED
            binding.watchedBoxEpisode.isEnabled = false
            binding.imageViewCollected.visibility = View.INVISIBLE
            binding.textViewEpisodeAlternativeNumbers.text = null
            binding.textViewEpisodeAirdate.text = null
            return
        }

        // episode title
        val watchedFlag = episode.watched
        val episodeNumber = episode.episodenumber
        val hideTitle = (EpisodeTools.isUnwatched(watchedFlag) && preventSpoilers(context))
        binding.textViewEpisodeTitle.text = TextTools.getEpisodeTitle(
            context, if (hideTitle) null else episode.title, episodeNumber
        )

        // number
        binding.textViewEpisodeNumber.text = integerFormat.format(episodeNumber.toLong())

        // watched box
        binding.watchedBoxEpisode.episodeFlag = watchedFlag
        binding.watchedBoxEpisode.isEnabled = true
        val watched = EpisodeTools.isWatched(watchedFlag)
        binding.watchedBoxEpisode.contentDescription =
            context.getString(if (watched) R.string.action_unwatched else R.string.action_watched)
        TooltipCompat.setTooltipText(
            binding.watchedBoxEpisode,
            binding.watchedBoxEpisode.context.getString(
                if (watched) R.string.action_unwatched else R.string.action_watched
            )
        )

        // collected tag
        val isCollected = episode.collected
        binding.imageViewCollected.visibility = if (isCollected) View.VISIBLE else View.INVISIBLE

        // alternative numbers
        val absoluteNumber = episode.absoluteNumber
        val absoluteNumberText = if (absoluteNumber > 0) {
            integerFormat.format(absoluteNumber.toLong())
        } else null
        val dvdNumber = episode.dvdNumber
        val dvdNumberText: String? = if (dvdNumber > 0) {
            "${context.getString(R.string.episode_number_disk)} $dvdNumber"
        } else null
        binding.textViewEpisodeAlternativeNumbers.text =
            TextTools.dotSeparate(absoluteNumberText, dvdNumberText)

        // release time
        val isReleased: Boolean
        val releaseTime = episode.firstReleasedMs
        if (releaseTime != -1L) {
            val actualRelease = TimeTools.applyUserOffset(context, releaseTime)
            isReleased = TimeTools.isReleased(actualRelease)
            // "in 15 mins" or "Oct 31, 2010"
            val displayExactDate = isDisplayExactDate(context)
            binding.textViewEpisodeAirdate.text =
                if (displayExactDate) TimeTools.formatToLocalDateShort(
                    context,
                    actualRelease
                ) else TimeTools.formatToLocalRelativeTime(context, actualRelease)
        } else {
            binding.textViewEpisodeAirdate.text =
                context.getString(R.string.episode_firstaired_unknown)
            isReleased = false
        }

        // dim text color if not released
        TextViewCompat.setTextAppearance(
            binding.textViewEpisodeTitle,
            if (isReleased) {
                R.style.TextAppearance_SeriesGuide_Subtitle1_Bold
            } else R.style.TextAppearance_SeriesGuide_Subtitle1_Bold_Dim
        )
        TextViewCompat.setTextAppearance(
            binding.textViewEpisodeAirdate,
            if (isReleased) {
                R.style.TextAppearance_SeriesGuide_Body2_Secondary
            } else R.style.TextAppearance_SeriesGuide_Body2_Dim
        )
    }

    companion object {
        fun create(
            parent: ViewGroup,
            clickListener: EpisodesAdapter.ClickListener
        ): EpisodeViewHolder {
            return EpisodeViewHolder(
                ItemEpisodeBinding.inflate(
                    LayoutInflater.from(parent.context),
                    parent,
                    false
                ),
                clickListener
            )
        }
    }
}
